### PR TITLE
client: Provide utilities for object creation

### PR DESCRIPTION
**Pull request by cthulhu-rider**
* relates to #342

### CODE REVIEW

**Comment by mialbu** [comment]
@cthulhu-rider Awesome! These methods are quite useful for us.

It would be awesome if you could provide similar functionality for reading an object and its header and also deleting an object.

Also, I have some other use case questions that don't necessarily fit into this PR, but they might:

1. Is there a simple way to just read an object's header, i.e., without a session?
2. It would be great to have a test that showcases how existing objects can be searched in a container. For example, if I create two objects and I forget about one, how can I know that there are two objects and get the "lost" one's object id in order to read its header and data/payload?
3. I've seen in https://github.com/nspcc-dev/neofs-aio that it's possible to give an object a filename in an example via nginx. How can I do that with `neofs-sdk-go`, and can I read an object also with just its filename instead of the object id? If it's possible to give a filename, am I correct that this would be shown when getting an object's header, or do I need to set this somehow in a custom manner?

**Comment by cthulhu-rider** [comment]
@mialbu I've added a new bunch of functions:
* read object
* remove object
* list all objects
* upload file
* download/restore file

1. You can do any operation w/o a session (may be except creation for now). Same for reading the header. Do you want to fetch some particular info from the header? I can add an example for it.
2. See added `ListObjects`
3. See added `UploadFile` / `RestoreFile`

**Comment by mialbu** [comment]
Hey @cthulhu-rider thanks for all these functions! 💯 

I've run into a problem with object ids.

I can write objects but somehow its ID is not set, so if I print the ID I get `11111111111111111111111111111111`, and if I write out `string(obj.ToV2().GetObjectID().GetValue())` (which is done internally in `neofs-sdk-go`) I get an empty string. However, when the code you provided runs and actually returns an `oid.ID` value, but when trying to get its header with curl it states `HTTP/1.1 301 Moved Permanently`, and when I try to read it with the read method you provided in this pr, I get an error of type `apistatus.ObjectNotFound` with message `status: code = 2049 message = object not found`.

Do I have to set this ID manually? If so, how can I do that?

**Comment by mialbu** [comment]
Hey @cthulhu-rider thanks for all these functions! 💯 

I've run into a problem with object ids.

I can write objects but somehow its ID is not set, so if I print the ID I get `11111111111111111111111111111111`, and if I write out `string(obj.ToV2().GetObjectID().GetValue())` (which is done internally in `neofs-sdk-go`) I get an empty string. However, when the code you provided runs and actually returns an `oid.ID` value, but when trying to get its header with curl it states `HTTP/1.1 301 Moved Permanently`, and when I try to read it with the read method you provided in this pr, I get an error of type `apistatus.ObjectNotFound` with message `status: code = 2049 message = object not found`.

Do I have to set this ID manually? If so, how can I do that?

File: client/clientutil/object.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+			attributes[i].SetValue(prm.attributes[i][1])
+		}
+
+		obj.SetAttributes(attributes...)
+	}
+
+	// write header
+	if streamObj.WriteHeader(obj) && prm.payload != nil { // do not swap the conditions!
+		var n int
+		buf := make([]byte, 100<<10)
```
Comment (roman-khimov) [inline client/clientutil/object.go:171]: Magic constant?

File: client/clientutil/object.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+// The object is stored in the parameterized container (required). Signer parameter
+// is used for object creation and communication over the NeoFS protocol (required).
+// There are some optional parameters, see CreateObjectPrm methods for details.
+//
+// Client connection MUST BE opened in advance, see Dial method for details.
+// Network communication is carried out within a given context, so it MUST NOT
+// be nil.
+//
+// See also CreateObject.
+func CreateObjectWithClient(ctx context.Context, c *client.Client, prm CreateObjectPrm) error {
```
Comment (roman-khimov) [inline client/clientutil/object.go:87]: How about making this `clntutil.CreateObject`...

File: client/clientutil/object.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+	return nil
+}
+
+// CreateObject creates new NeoFS object and stores it into the NeoFS network
+// through the given endpoint.
+//
+// CreateObject is well suited for one-time data storage. To create multiple
+// objects using the same endpoint, use CreateObjectWithClient.
+func CreateObject(ctx context.Context, endpoint string, prm CreateObjectPrm) error {
```
Comment (roman-khimov) [inline client/clientutil/object.go:210]: ... and this `clntutil.ConnectAndCreateObject`? I'd expect this one to be used a bit less frequently (I might be wrong with this assumption though).

File: client/clientutil/object.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
+package clientutil
```
Comment (roman-khimov) [inline client/clientutil/object.go:1]: I'm not sure this is worth another package.

File: client/clientutil/init.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"fmt"
+)
+
+// global default signer.
+var signerDefault ecdsa.PrivateKey
```
Comment (roman-khimov) [inline client/clientutil/init.go:11]: This looks dangerous. Keys must be managed explicitly.

File: client/clientutil/object.go [hunk 6]
```diff
@@ -202,19 +203,151 @@ func CreateObjectWithClient(ctx context.Context, c *client.Client, prm CreateObj
...
+type ReadObjectPrm struct {
+	// Target NeoFS container.
+	Container cid.ID
+
+	// Object's reference to read from the Container.
+	Object oid.ID
 
-	var c client.Client
-	c.Init(prmInit)
+	payloadWriter io.Writer
```
Comment (roman-khimov) [inline client/clientutil/object.go:227]: Why unexported here with exported CID/OID above?

File: client/clientutil/object.go [hunk 7]
```diff
@@ -202,19 +203,151 @@ func CreateObjectWithClient(ctx context.Context, c *client.Client, prm CreateObj
...
+
+	return nil
+}
+
+// ReadObject reads object from the NeoFS network through the given endpoint.
+//
+// ReadObject is well suited for one-time data reading. To read multiple
+// objects using the same endpoint, use ReadObjectWithClient. ReadObject
+// inherits behavior of ReadObjectWithClient.
+func ReadObject(ctx context.Context, endpoint string, prm ReadObjectPrm) error {
```
Comment (roman-khimov) [inline client/clientutil/object.go:295]: Maybe these could be some `quick` API, while the other are more like Client (or upper-level thing) methods.

**Comment by cthulhu-rider** [comment]
refs #380

**Comment by cthulhu-rider** [comment]
seems like no longer needed with https://github.com/nspcc-dev/neofs-sdk-go/releases/tag/v1.0.0-rc.10

**Pull request closed without merging**